resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -867,6 +868,20 @@ object ApplicationMaster extends Logging {
         val originalCreds = UserGroupInformation.getCurrentUser().getCredentials()
         SparkHadoopUtil.get.loginUserFromKeytab(principal, sparkConf.get(KEYTAB).orNull)
         val newUGI = UserGroupInformation.getCurrentUser()
+
+        if (master.isClusterMode) {
+          // Set the context class loader so that the token manager has access to jars
+          // distributed by the user.
+          Utils.withContextClassLoader(master.userClassLoader) {
+            // Re-obtain delegation tokens, as they might be outdated as of now. Add the fresh
+            // tokens on top of the original user's credentials (overwrite).
+            // This is only needed in cluster mode, because in client mode, AM will soon retrieve
+            // the latest tokens from the driver.
+            val credentialManager = new HadoopDelegationTokenManager(sparkConf, yarnConf, null)
```

Comment:
does this already handle the login from key tab that is done above? at a glance it looks like it should by doing the doLogin() call.

Reply:
https://github.com/apache/spark/blob/bd264299317bba91f2dc1dc27fd51e6bc0609d66/core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala#L145

Reply:
right so my question is can we get rid of line 869 above where it         SparkHadoopUtil.get.loginUserFromKeytab(principal, sparkConf.get(KEYTAB).orNull)
 if it is already done inside obtainDelegationTokens.  it looks like you can from the code but I would want to test to verify.

Reply:
The section `Why are the changes needed?` in the description of the PR covers it, so please go through it. We "intentionally" obtain delegation tokens here, though driver will do it again, because driver (a.k.a user code) "before" initializing SparkContext may require delegation tokens.

The sample application demonstrates the issue clearly:

https://github.com/HeartSaVioR/spark-delegation-token-experiment/blob/master/src/main/scala/net/heartsavior/spark/example/LongRunningAppWithHDFSConfig.scala

Reply:
My previous comment referred incorrect line. My bad. Just deleted.

Looks like `HadoopDelegationTokenManager.obtainDelegationTokens` checks the current UGI which is changing by the line 869. `doLogin()` doesn't change the current UGI so that's the difference.